modules/apps/frontend-js/frontend-js-components-web/src/main/resources/META-INF/resources/walkthrough/observeRect.js

```diff
@@ -0,0 +1,50 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+const rectAttrs = ['bottom', 'height', 'left', 'right', 'top', 'width'];
+
+const rectChanged = (a, b) => rectAttrs.some((prop) => a[prop] !== b[prop]);
+
+let rafId;
+
+const run = (node, state) => {
+	const newRect = node.getBoundingClientRect();
+
+	if (rectChanged(newRect, state.rect)) {
+		state.rect = newRect;
+
+		state.callback(state.rect);
+	}
+
+	rafId = window.requestAnimationFrame(() => run(node, state));
```

Comment:
> Isn't this an endless loop? scream

No, this code will run inside a requestAnimationFrame and it will not be an infinite loop

> What is the purpose of this whole observeRect logic?

Deal with resizings, DOM modifications. It was originally created on Clay to solve some issues like Dropdown Menu positioning when We resize things. See the following video:


https://user-images.githubusercontent.com/7663513/167028401-77127755-88a1-42a8-b3e2-40628ddf3c35.mp4

Reply:
...Which is not working,

Thanks for finding on https://github.com/liferay-frontend/liferay-portal/pull/2164#issuecomment-1118762792

Reply:
I added a console log line to illustrate the loop:

![screenshot](https://user-images.githubusercontent.com/5435511/167155421-45f41b71-aa51-4dea-80aa-c82fb91a83bf.gif)

Reply:
Regarding updating many times states of a React component and how it handles it I think you can dig deeper into the React architecture explanations, I would recommend reading [React Fiber Architecture](https://github.com/acdlite/react-fiber-architecture) this is a look at how to React deals with renderings and updating state this is quite old and it changed a little with the React 18 and the introduction of [React Concurrent](https://reactjs.org/blog/2022/03/29/react-v18.html#what-is-concurrent-react), on the React blog talks a little about it a little more if you want to understand more about it how things work under the hood I would recommend reading some articles from the [React 18 working group](https://github.com/reactwg/react-18/discussions), they are great information.

Reply:
Thanks for clarifications, and recommended reads! :heart:  I'll make sure to take a look.

> Basically, we want to know when the element changes position, and we need it to be called between 30~60 times to guarantee the FPS and that it doesn't slow down, otherwise the user will notice that the element starts to move slowly and it feels like if the page is slow

Yeah, that was always the downside of debouncing. It was a necessary measure though, as something like a manual window resize causes continuous element position changes. I've seen several cases in my time where a resize would completely clog JS execution. But, it looks like React automagically handles this. Awesome!